app/controllers/class_lists_controller.rb

```diff
@@ -161,6 +160,32 @@ def update_class_list_json
     }
   end
 
+  # This is a POST for storing all principal revisions.
+  def principal_revised_class_list_json
+    params.require(:workspace_id)
+    params.require(:principal_revisions_json)
+    workspace_id = params[:workspace_id]
+    principal_revisions_json = params[:principal_revisions_json]
+
+    # The class list must be submitted already
+    latest_class_list = ClassList.latest_class_list_for_workspace(workspace_id)
+    raise Exceptions::EducatorNotAuthorized unless latest_class_list.submitted?
+
+    # They must be authorized to read it
+    grade_level_now = GradeLevels.new.previous(latest_class_list.grade_level_next_year)
+    raise Exceptions::EducatorNotAuthorized unless queries.is_authorized_for_grade_level_now?(latest_class_list.school_id, grade_level_now)
+
+    # They must be authorized to revise it
+    raise Exceptions::EducatorNotAuthorized unless queries.is_authorized_to_revise?(latest_class_list)
+
+    # Write a new record
+    class_list = latest_class_list.dup
```

Comment:
Ooh, this is neat! I wasn't familiar with ActiveRecord `.dup`, dug in a little bit here: https://apidock.com/rails/ActiveRecord/Core/dup.

One thing that wasn't immediately obvious to me here: do we save the principal's educator ID anywhere on the `class_list` record here? How can we trace a set of class list revisions back to the principal who made them?

Reply:
@alexsoble thanks, this is a great question!  Right now, each change any user makes is stored as a separate `class_list` record.  They each have a `workspace_id` field, which describes a logical "4th grade class last that teacher Sarah started."

So if you do:

```
ClassList.where(workspace_id: 'foo')
```

You'll get a bunch of records that show the changes over time, including the teacher edits and then later the principal edits.  The `created_by_educator_id` would always refer to the original teacher.

This PR didn't save the principal's id anywhere, I updated that now.  I also renamed `created_by_educator_id` to `created_by_teacher_educator_id` to make it a little less foot-gun, since the created by field usually means who created the Active Record model.

Overall, this is all a consequence of putting things in one table to track all the changes over time. :)  A good improvement here would be to migrate everything from one table into three table: `workspaces`, `teacher_edits` and `principal_revisions`.  That would make the validations simpler and express more explicitly what's happening here.  For now I added more explanation around this here: https://github.com/studentinsights/studentinsights/pull/1746/files#diff-8130f19224e55dd36a1da7f6ae8e5ddeR1